### PR TITLE
Add collaboration related guidelines

### PR DESCRIPTION
**Pull request by harendra-kumar**
Also, add a section on searching historical commits.

### CODE REVIEW

File: git-intro.rst [hunk 1]
```diff
@@ -193,38 +200,93 @@ repository set as a remote named `upstream`. When you clone your fork
...
+~~~~~~~~~~~~~~~~
+
+* Do not directly push changes to the master branch or for that matter
+  to to any collaboration branch. Collaboration branches should be pushed
+  only when merging.
+* Never force push on collaboration branches. They always go in the
+  forward direction, they are never rebased or reset to previous points
+  except in very unusual cases.  If something has to be reverted on these
+  branches it has to be committed as another forward commit. A forced change on
+  collaboration branched can create havoc as everyone working on the branch
```
Comment (specdrake) [inline git-intro.rst]: collaboration `branches`

File: git-intro.rst [hunk 2]
```diff
@@ -193,38 +200,93 @@ repository set as a remote named `upstream`. When you clone your fork
...
+
+Once the review is done you can rebase/squash your changes and force
+push.
 
-* git rebase -i
+Engagement rules
+~~~~~~~~~~~~~~~~
+
+* Do not directly push changes to the master branch or for that matter
+  to to any collaboration branch. Collaboration branches should be pushed
```
Comment (specdrake) [inline git-intro.rst]: extra `to`

File: git-intro.rst (hunk 1 above)
Comment (harendra-kumar) [reply git-intro.rst]: fixed.

File: git-intro.rst (hunk 2 above)
Comment (harendra-kumar) [reply git-intro.rst]: fixed.

**Pull request merged**